fix: removes extra domains from `SelectedNetworkController` domain state

adonesky1 opened the pull request:
- Bumps to the the latest version of `@metamask/selected-network-controller` which fixes issue where the `SelectedNetworkController` adds any and all domains the user visits to its `domains` state whether or not the user has connected to these sites. 
- Adds a migration to clean up the undesired state caused by this bug.
- Adds logic to listen for updates to permissions state and add newly connected/permissioned dapps to `domains` state in the `SelectedNetworkController`. I've created a ticket [here](https://app.zenhub.com/workspaces/wallet-api-platform-63bee08a4e3b9d001108416e/issues/gh/metamask/metamask-planning/2142) to move this logic into the `SelectedNetworkController` itself.
- Adds logic to couple the `useRequestQueue` feature flag with the `perDomainNetwork` feature flag. I've created a ticket [here](https://app.zenhub.com/workspaces/wallet-api-platform-63bee08a4e3b9d001108416e/issues/gh/metamask/metamask-planning/2143) to remove the latter state since it is redundant and requires keeping these two booleans in sync.
- Removes calls to `setNetworkClientIdForMetamask` since the [latest version of `@metamask/selected-network-controller`](https://github.com/MetaMask/core/blob/main/packages/selected-network-controller/CHANGELOG.md#800) removes this method (and the `metamask` domain from the `domains` state entirely).
- Adds logic to use provider and blockTracker from the `SelectedNetworkController` when feature flag is on and connected origin is in `domains` state (which can only happen when the dapp is connected) in `setupProviderEngine` for RPC middleware stack.
- Fixes a bug where if a dapp prompts an `addEthereumChain` request and the user then accepts the subsequent `switchEthereumChain` call that is chained on, the `networkClientID` set for that dapp in the `SelectedNetworkController` domains state will not be updated.

Manual Testing Steps

well first you need to [comment out some code](https://github.com/MetaMask/metamask-extension/pull/22964) that is causing [a separate bug (found on develop)](https://github.com/MetaMask/metamask-extension/issues/22937) but one that will make manually testing this a nightmare.

1. Go to `settings` > `experimental` > `Select networks for each site` and toggle this setting on
2. Put the globally selected network to ethereum mainnet
3. Connect to [uniswap](https://app.uniswap.org/swap)
4. Connect to [pancakeswap](https://pancakeswap.finance/)
5. Vist [Aave](https://app.aave.com/) but don't connect the wallet
6. Open devtools in the wallet UI and run the follow command in the console:
```
const { metamask } = await window.stateHooks.getCleanAppState();
console.log(metamask.domains)
```
7. You should see the following log:
```
{https://app.uniswap.org: 'mainnet', https://pancakeswap.finance: 'mainnet'}
```
8. Now call `switchEthereumChain` on one of the connected dapps (they each have UI elements that do this)
9. Now go back to the other dapp (where the chain hasn't been switched) and initiate a transaction
10. You should receive a `switchEthereumChain` confirmation back to Ethereum Mainnet before the transaction confirmation is show.

adonesky1 committed:
Bump @metamask/selected-network-controller to v8.0.0, fix issue where all visted domains are added to SelectedNetworkController state, use migration to clean up state

adonesky1 commented on app/scripts/lib/rpc-method-middleware/handlers/switch-ethereum-chain.js:
```diff
@@ -26,7 +26,6 @@ const switchEthereumChain = {
     setActiveNetwork: true,
     requestUserApproval: true,
     getNetworkConfigurations: true,
-    getNetworkClientIdForDomain: true,
```
this wasn't being used?

adonesky1 commented on app/scripts/metamask-controller.js:
```diff
@@ -493,11 +494,31 @@ export default class MetamaskController extends EventEmitter {
+      state: initState.SelectedNetworkController,
     });
 
+    // If the perDomainNetwork feature flag is enabled, use metamask origin provider and block tracker
+    // as the proxy passed to the controllers instantiated below, otherwise use the network controller's provider and block tracker
+    // let networkClient;
+    // if (this.selectedNetworkController.state.perDomainNetwork) {
+    //   networkClient =
+    //     this.selectedNetworkController.getProviderAndBlockTracker(
+    //       METAMASK_DOMAIN,
+    //     );
+    // } else {
+    // }
```
for testing purposes due to https://github.com/MetaMask/metamask-extension/issues/22937 / https://github.com/MetaMask/metamask-extension/pull/22964

adonesky1 commented on ui/components/ui/new-network-info/new-network-info.js:
```diff
@@ -42,29 +42,29 @@ export default function NewNetworkInfo() {
-    // we want to only fetch once
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, []);
+  // useEffect(() => {
+  //   checkTokenDetection();
+  //   // we want to only fetch once
+  //   // eslint-disable-next-line react-hooks/exhaustive-deps
+  // }, []);
```
for testing purposes due to https://github.com/MetaMask/metamask-extension/issues/22937 / https://github.com/MetaMask/metamask-extension/pull/22964

adonesky1 commented on app/scripts/metamask-controller.js:
```diff
@@ -4913,10 +4940,6 @@ export default class MetamaskController extends EventEmitter {
             this.networkController,
           ),
         setActiveNetwork: (networkClientId) => {
-          // setActiveNetwork is used for custom (non-infura) networks, and for these networks networkClientId === networkConfigurationId
-          this.selectedNetworkController.setNetworkClientIdForMetamask(
-            networkClientId,
-          );
```
this is no longer necessary because `SelectedNetworkController` no longer keeps track of the globally `selectedNetworkClientId`

adonesky1 commented on app/scripts/metamask-controller.js:
```diff
@@ -4938,8 +4961,6 @@ export default class MetamaskController extends EventEmitter {
         ),
         getProviderConfig: () => this.networkController.state.providerConfig,
         setProviderType: (type) => {
-          // when using this format, type happens to be the same as the networkClientId...
-          this.selectedNetworkController.setNetworkClientIdForMetamask(type);
```
this is no longer necessary because `SelectedNetworkController` no longer keeps track of the globally `selectedNetworkClientId`

adonesky1 commented:
@metamaskbot update-policies

adonesky1 commented on app/scripts/metamask-controller.js:
```diff
@@ -2417,6 +2442,17 @@ export default class MetamaskController extends EventEmitter {
+              this.networkController.state.selectedNetworkClientId,
+            );
+          } else {
+            // TODO remove domain from selectedNetworkController when permissions are removed?
```
I don't think theres any harm in leaving this state as if the user ever reconnects to this domain it will hit the logic right above and reset the selectedNetworkClientId to the globally selected network but probably cleaner to add a method to remove a domain from state on the SelectedNetworkController. wdyt @BelfordZ

BelfordZ replied:
yeah I think we are good to go without it, at least I cant think of any bug that would arise from leaving the state as it is.

adonesky1 commented on app/scripts/metamask-controller.js:
```diff
@@ -493,11 +494,34 @@ export default class MetamaskController extends EventEmitter {
+    });
+
+    // couples the useRequestQueue featureflag with the perDomainNetwork feature flag
+    this.preferencesController.store.subscribe(({ useRequestQueue }) => {
+      if (
+        useRequestQueue !==
+        this.selectedNetworkController.state.perDomainNetwork
+      ) {
+        this.selectedNetworkController.setPerDomainNetwork(useRequestQueue);
+      }
```
we should probably just remove this redundant `perDomainNetwork` state from the `SelectedNetworkController` and just query for the `useRequestQueue` state on the `PreferencesController` wherever we need it. But this will require another update to the `SelectedNetworkController` and should be done in a separate PR.

adonesky1 replied:
This dedupe is happening here btw: https://github.com/MetaMask/core/pull/3989

jiexi commented on app/scripts/metamask-controller.js:
```diff
@@ -2417,6 +2441,17 @@ export default class MetamaskController extends EventEmitter {
 
         for (const [origin, accounts] of changedAccounts.entries()) {
           this._notifyAccountsChange(origin, accounts);
+          // When permissions are added for a given origin we need to
+          // set the network client id for that origin to the current
+          // globally selected network client id.
+          if (accounts.length > 0) {
+            this.selectedNetworkController.setNetworkClientIdForDomain(
+              origin,
+              this.networkController.state.selectedNetworkClientId,
+            );
```
this is only necessary because we access domain state directly, right? because `getNetworkClientIdForDomain()` would already return the selectedNetworkClientId as a fallback

adonesky1 replied:
So this is actually the point at which we want to add the newly connected dapp to the `domains` state in `SelectedNetworkController`. 

However, this is another candidate for something that can/should be pulled directly into the `SelectedNetworkController`

jiexi replied:
Sounds good. the less MetaMaskController has to be concerned about, the better

adonesky1 commented on app/scripts/lib/setupSentry.js:
```diff
@@ -225,7 +225,7 @@ export const SENTRY_BACKGROUND_STATE = {
     useRequestQueue: true,
   },
-  SelectedNetworkController: { domains: true, perDomainNetwork: false },
+  SelectedNetworkController: { domains: false, perDomainNetwork: true },
```
We do not want to include domains in our sentry reports since it is potentially sensitive and violates our goal to preserve privacy, but we do want to include the perDomainNetwork toggle since it is not sensitive and potentially useful for diagnosing bugs.

adonesky1 committed:
add comments

adonesky1 commented on app/scripts/metamask-controller.js:
```diff
@@ -2422,6 +2446,19 @@ export default class MetamaskController extends EventEmitter {
 
         for (const [origin, accounts] of changedAccounts.entries()) {
           this._notifyAccountsChange(origin, accounts);
+          // When permissions are added for a given origin we need to
+          // set the network client id for that origin to the current
+          // globally selected network client id.
+          if (accounts.length > 0) {
+            // TODO add this logic directly in the SelectedNetworkController
+            this.selectedNetworkController.setNetworkClientIdForDomain(
+              origin,
+              this.networkController.state.selectedNetworkClientId,
+            );
+          } else {
+            // TODO remove domain from selectedNetworkController when permissions are removed
+            // Add this logic directly in the SelectedNetworkController
+          }
```
There is a WIP PR against the `SelectedNetworkController` to pull this logic into that controller instead: https://github.com/MetaMask/core/pull/3969

tmashuang committed:
test: Bump selected network controller v8.0.0 fix test (#23190)

## **Description**

Updates `multiple-networks-dapps-txs.spec.js` to connect, switch
networks, and send txs with requesting changes.

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/23190?quickstart=1)

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it
is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
- [ ] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

jiexi commented:
tested, working

jiexi commented on app/scripts/metamask-controller.js:
```diff
@@ -520,6 +530,20 @@ export default class MetamaskController extends EventEmitter {
         ),
     });
 
+    // Couple the useRequestQueue featureflag with the perDomainNetwork feature flag
```
```suggestion
    // Couple the useRequestQueue feature flag with the perDomainNetwork feature flag
```

nit

jiexi commented on app/scripts/metamask-controller.js:
```diff
@@ -2609,7 +2646,7 @@ export default class MetamaskController extends EventEmitter {
       accounts: await this.getPermittedAccounts(origin),
       ...this.getProviderNetworkState(
-        this.preferencesController.getUseRequestQueue() ? origin : undefined,
+        this.preferencesController?.getUseRequestQueue() ? origin : undefined,
```
why is this needed now?

adonesky1 replied:
not seeing this issue anymore so removing this for now: https://github.com/MetaMask/metamask-extension/commit/f2c2dbdb049f2e7268be45d684c8714acc1833f9

jiexi commented on app/scripts/metamask-controller.js:
```diff
@@ -4728,25 +4760,21 @@ export default class MetamaskController extends EventEmitter {
-        this.selectedNetworkController.getProviderAndBlockTracker(
-          origin,
-        ).provider;
+    // Use proxyClient from selectedNetworkController when:
+    // 1. feature flag for perDomainNetwork is on
+    // 2  the origin is in the selectedNetworkController's `domains` state
```
I think it may be more helpful to future devs to explain why we prefer using the global as opposed to setting the networkClient for the origin if it doesn't exist already and then using that. I could see someone trying to set a default here and not having enough background/context to understand why we don't want to do that

adonesky1 replied:
Chained the slant of the comments here: https://github.com/MetaMask/metamask-extension/pull/23130/commits/f2c2dbdb049f2e7268be45d684c8714acc1833f9 . Let me know what you think

jiexi commented on test/e2e/webdriver/driver.js:
```diff
@@ -546,6 +546,43 @@ class Driver {
+      await this.delay(delayStep);
+      timeElapsed += delayStep;
+      // refresh the window handles
+      windowHandles = await this.driver.getAllWindowHandles();
```
is it possible to do this all within the `await retry`?

```
await retry(
          {
            retries,
            delay: retryDelay,
          },
          async () => {
            windowHandles = await this.driver.getAllWindowHandles();
            for (const handle of windowHandles) {
              await this.driver.switchTo().window(handle);
              const handleUrl await this.driver.getCurrentUrl();
              if (handleUrl === `${url}/`) {
                return true;
              }
            }
          },
        );

```

jiexi replied:
something like that?

adonesky1 replied:
cc @tmashuang ☝️

tmashuang replied:
I tried what @jiexi proposed and It didn't seem to work. Looks like it wasn't able to break out of the `retry()`

jiexi replied:
ahhh, i'm sorry. Thank you for taking on the burden of trying it. Apologies 🙏

adonesky1 committed:
add a patch to add any existing permissioned domains to the selected network controller domains if missing from that state.

adonesky1 committed:
fix bug with add-ethereum-chain

adonesky1 committed:
lint?

adonesky1 commented on app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:
```diff
@@ -294,6 +296,7 @@ async function addEthereumChainHandler(
         fromNetworkConfiguration: getProviderConfig(),
       },
     });
+    setNetworkClientIdForDomain(req.origin, networkConfigurationId);
```
This fixes a bug where if a dapp prompts an add-ethereum-chain request and the user then accepts the subsequent `switchEthereumChain` call that is chained on, the `networkClientID` set for that dapp in the `SelectedNetworkController` domains state will not be updated.

adonesky1 committed:
fix addEthereumChain handler so that when chained switchEthereumChain call is approved the SelectedNetworkController state is updated correct

darkwing commented:
I can confirm this PR fixes my `domains` key/value issue!

shanejonas commented on app/scripts/metamask-controller.js:
```diff
@@ -2620,9 +2673,12 @@ export default class MetamaskController extends EventEmitter {
     let chainId;
-    if (this.preferencesController.getUseRequestQueue()) {
+    if (
+      this.preferencesController?.getUseRequestQueue() &&
```
```suggestion
      this.preferencesController.getUseRequestQueue() &&
```
`this.preferencesController` should always exist

adonesky1 replied:
unless, as appears to be the case here in some scenarios, this method is called before the controller is instantiated.

adonesky1 replied:
not seeing this issue anymore so removing this for now: https://github.com/MetaMask/metamask-extension/pull/23130/commits/f2c2dbdb049f2e7268be45d684c8714acc1833f9

Gudahtt commented on app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:
```diff
@@ -294,6 +298,9 @@ async function addEthereumChainHandler(
       },
     });
+    if (hasPermission(req.origin)) {
+      setNetworkClientIdForDomain(req.origin, networkConfigurationId);
```
This solution seems OK, but it makes me wonder if similar cases exist elsewhere.

Have we considered listening to permission state changes, and automatically setting a selected chain when a new origin is granted permissions? That would ensure that we never "forget" to set this initial selected network client.

adonesky1 replied:
> Have we considered listening to permission state changes, and automatically setting a selected chain when a new origin is granted permissions

Yes indeed. In fact we're doing that in this PR. See [here](https://github.com/MetaMask/metamask-extension/pull/23130/files#r1503052895).

The problem here is not the **initial** network being set, its that we need to change the networkClientId for a dapp that already is permissioned.

Gudahtt replied:
Ahh I see,  I follow now, thanks

adonesky1 committed:
Update app/scripts/metamask-controller.js

Co-authored-by: jiexi <jiexiluan@gmail.com>

adonesky1 committed:
modify comments and remove optional chaining

jiexi commented on test/e2e/webdriver/driver.js:
```diff
@@ -546,6 +546,43 @@ class Driver {
+
+    while (timeElapsed <= timeout) {
+      for (const handle of windowHandles) {
+        const handleUrl = await retry(
+          {
+            retries,
+            delay: retryDelay,
+          },
+          async () => {
+            await this.driver.switchTo().window(handle);
+            return await this.driver.getCurrentUrl();
+          },
+        );
+
+        if (handleUrl === `${url}/`) {
+          return handle;
+        }
+      }
```
ahh i'm sorry to comment on this one again. It just seems weird to have what is essentially two retry/timeout loops. Is this inner `retry` necessary? 

timeElapsed isn't really the time elapsed since it doesn't account for how long the execution of the loop took

tmashuang replied:
I believe the retry is necessary, maybe the `while (timeElapsed <= timeout)` part might not be though. These methods have been combined to combat the flaky tests [PR](https://github.com/MetaMask/metamask-extension/pull/21394) and internal discussions as to the best approach [internal disc](https://consensys.slack.com/archives/GTQAGKY5V/p1697223609357109?thread_ts=1697150920.731179&cid=GTQAGKY5V)  and an open issue https://github.com/MetaMask/metamask-extension/issues/21403 for future refactoring in the future.

tmashuang replied:
Thinking more largely, we could combine all the switching windows methods together `switchToWindow`, `switchToWindowWithTitle`, and `switchToWindowWithUrl`.

jiexi replied:
Thank you for explaining 🙏

adonesky1 commented on app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:
```diff
@@ -294,6 +298,9 @@ async function addEthereumChainHandler(
       },
     });
+    if (hasPermission(req.origin)) {
+      setNetworkClientIdForDomain(req.origin, networkConfigurationId);
```
I wonder if we can also add this directly in the `SelectedNetworkController` by listening for accepted `switchEthereumChain` requests from the `ApprovalController`? wdyt @Gudahtt @mcmire @BelfordZ

adonesky1 replied:
Looked into this doesn't look very straightforward unfortunately.

adonesky1 committed:
Merge branch 'develop' into ad/bump-selected-network-controller-v8.0.0

adonesky1 committed:
Merge branch 'develop' into ad/bump-selected-network-controller-v8.0.0

adonesky1 committed:
Merge branch 'develop' into ad/bump-selected-network-controller-v8.0.0

Gudahtt commented on app/scripts/metamask-controller.js:
```diff
@@ -2618,9 +2671,12 @@ export default class MetamaskController extends EventEmitter {
+  getProviderNetworkState(origin = METAMASK_DOMAIN) {
     let chainId;
-    if (this.preferencesController.getUseRequestQueue()) {
+    if (
```
Just noticed a bug in this function: the `networkVersion` returned is always for the globally selected network, even when this feature is on. That would result in the provider having a `networkId` that contradicts the `chainId`. Not good.

Not a blocker for this PR since it wasn't introduced here, but we should fix this as well.

adonesky1 replied:
Heres a ticket: https://github.com/MetaMask/MetaMask-planning/issues/2176

Gudahtt commented on app/scripts/metamask-controller.js:
```diff
@@ -5767,7 +5809,7 @@ export default class MetamaskController extends EventEmitter {
 
   _notifyChainChange() {
-    if (this.preferencesController.getUseRequestQueue()) {
+    if (this.preferencesController?.getUseRequestQueue()) {
```
Do we still need this change?

Maybe similar to [this case](https://github.com/MetaMask/metamask-extension/pull/23130#discussion_r1504540531), in that it's no longer needed

adonesky1 replied:
Reverted here: https://github.com/MetaMask/metamask-extension/pull/23130/commits/9af40f307ef4dcb2536aee62a5aef6ebb3126f6e

Gudahtt commented on app/scripts/migrations/111.ts:
```diff
@@ -0,0 +1,44 @@
+function transformState(state: Record<string, any>) {
+  if (
+    !hasProperty(state, 'SelectedNetworkController') ||
+    !isObject(state.SelectedNetworkController) ||
```
Nit: In other migrations we distinguished between the "Missing state that is expected to be missing" scenarios, and "Malformed/corrupted state".

We would expect the `SelectedNetworkController` state to be missing if the user upgraded from an older wallet, so returning `state` unchanged for that scenario seems good. But if that state exists, the remaining three conditions should be true for all users unless the state was somehow corrupted. So we could return an error in those scenarios rather than returning state unchanged, to alert us to that state corruption.

adonesky1 replied:
added error logging here: https://github.com/MetaMask/metamask-extension/pull/23130/commits/bae4ca69e0077ef03f7c781c066cd31d3698e7e0

Gudahtt commented on app/scripts/lib/rpc-method-middleware/handlers/switch-ethereum-chain.js:
I'm seeing this change entry in v8 of the SelectedNetworkController:

>`setNetworkClientIdForDomain` now fails and throws an error if the passed in `domain` is not currently permissioned in the `PermissionsController`

yet this middleware will call `setNetworkClientIdForDomain` irrespective of whether the domain has permissions or not

adonesky1 replied:
Good call! This is not a flow I thought of, but is technically possible as well: where a dapp does not connect but prompts the wallet to switch chains. Will add permission check here as well!

adonesky1 replied:
added here: https://github.com/MetaMask/metamask-extension/pull/23130/commits/7cbb194fe92181d0c2bfe8eb9f2be45de4d54ef3

Gudahtt commented on test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js:
```diff
@@ -23,8 +25,8 @@ describe('Request Queuing for Multiple Dapps and Txs on different networks.', fu
-          .withPermissionControllerConnectedToTwoTestDapps()
-          .withSelectedNetworkControllerPerDomain()
+          // .withPermissionControllerConnectedToTwoTestDapps()
+          // .withSelectedNetworkControllerPerDomain() // TODO: Add back when refreshing state doesn't revert per dapp selected network. Remove having to connect to dapp
```
Hmm, what does this comment refer to? I'm not familiar with this problem

adonesky1 replied:
I'm not totally understanding this either. Currently working with Thomas to figure it out!

adonesky1 replied:
ok this was a misunderstanding. I worked through it with @tmashuang  and we've updated the e2e tests here https://github.com/MetaMask/metamask-extension/pull/23130/commits/85a1d28989f7f61472c0ad65a787a98023957752 so that they work as expected now. Let me know if you have any further questions @Gudahtt

adonesky1 committed:
address feedback/cleanup

adonesky1 commented on app/scripts/metamask-controller.js:
```diff
@@ -4726,25 +4774,25 @@ export default class MetamaskController extends EventEmitter {
+      // this means that origin does not have permissions (is not connected to the wallet)
+      // and will therefore not have its own selected network even if perDomainNetwork is true
+      // and so in this case too we want to use the globally selected network provider/blockTracker
+      proxyClient = {
```
Done here: https://github.com/MetaMask/metamask-extension/pull/23130/commits/9af40f307ef4dcb2536aee62a5aef6ebb3126f6e

adonesky1 committed:
add permission check in switchEthereumChain handler

Gudahtt commented on app/scripts/metamask-controller.js:
```diff
@@ -492,11 +493,20 @@ export default class MetamaskController extends EventEmitter {
+      state: initState.SelectedNetworkController,
     });
 
+    const networkClient = this.networkController.getProviderAndBlockTracker();
+    this.provider = networkClient.provider;
+    this.blockTracker = networkClient.blockTracker;
```
Following 9af40f307ef4dcb2536aee62a5aef6ebb3126f6e, this seems to no longer be used

```suggestion
```

adonesky1 replied:
removed here: https://github.com/MetaMask/metamask-extension/pull/23130/commits/bf5bc3ed7b7f769934f626e0752342261f87ba70

Gudahtt commented on ui/components/ui/new-network-info/new-network-info.js:
```diff
@@ -61,7 +61,7 @@ export default function NewNetworkInfo() {
 
   useEffect(() => {
-    checkTokenDetection();
+    // checkTokenDetection();
```
Was this change intentional?

adonesky1 replied:
nope! reverted here: https://github.com/MetaMask/metamask-extension/pull/23130/commits/0e8ee1c6417c0d4bbb0be7222bd03e27d5a5d7ab

tmashuang commented:
Some issues I ran into with my testing of the setting would be that there wasn't a good way to clear all the domains when debugging. It can be cleared individually via the connected dapp modal, but I would hope that the toggle would clear the `domain` state when turned off. Reiterating the issue with state refreshing not persisting the network per domain again as every update to the extension will revert the behavior. Also, the persisting `eth_chainId` sent back to the dapp even though the setting is toggled on.

adonesky1 committed:
add error logging for malformed state

adonesky1 committed:
revert unintentional commenting out

adonesky1 committed:
revert redundant provider and blocktracker setting

adonesky1 committed:
fix: hasPermission -> hasPermissions

adonesky1 committed:
Merge branch 'develop' into ad/bump-selected-network-controller-v8.0.0

adonesky1 committed:
fix e2e test

tmashuang commented on test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js:
```diff
@@ -85,25 +85,48 @@ describe('Request Queuing for Multiple Dapps and Txs on different networks.', fu
-        const windowHandles = await driver.getAllWindowHandles();
-        const dappOne = windowHandles[1];
-        const dappTwo = windowHandles[2];
+        // // Connect to dapp
```
```suggestion
        // Connect to dapp
```

tmashuang commented on test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js:
```diff
@@ -40,34 +42,32 @@ describe('Request Queuing for Multiple Dapps and Txs on different networks.', fu
+
+        await driver.delay(regularDelayMs);
+
+        // // Connect to Dapp
```
```suggestion
        // Connect to Dapp
```

adonesky1 committed:
cleanup

Gudahtt commented:
The diff looks good! I can re-review when CI passes. It looks like at least one of the CI failures is related to these changes, the e2e test failure.

adonesky1 committed:
Merge branch 'develop' into ad/bump-selected-network-controller-v8.0.0

adonesky1 committed:
Merge branch 'develop' into ad/bump-selected-network-controller-v8.0.0

jiexi commented on test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js:
```diff
@@ -40,34 +42,32 @@ describe('Request Queuing for Multiple Dapps and Txs on different networks.', fu
+
+        await driver.delay(regularDelayMs);
+
+        // Connect to Dapp
```
```suggestion
```

this comment already exists above on line 51

jiexi commented on test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js:
```diff
@@ -85,25 +85,48 @@ describe('Request Queuing for Multiple Dapps and Txs on different networks.', fu
+
+        await driver.delay(regularDelayMs);
+
+        // Connect to Dapp
```
```suggestion
```

adonesky1 committed:
remove unnecessary comments

tmashuang committed:
Test improvements

tmashuang committed:
Remove debugging

tmashuang committed:
Merge branch 'develop' into ad/bump-selected-network-controller-v8.0.0

Gudahtt reviewed (approved):
LGTM!
